server/core_storage.go

```diff
@@ -681,16 +682,15 @@ func storagePrepBatch(batch *pgx.Batch, authoritativeWrite bool, op *StorageOpWr
 			ON CONFLICT (collection, key, user_id) DO
 				UPDATE SET value = $4, version = $5, read = $6, write = $7, update_time = now()
 				WHERE TRUE` + writeCheck + `
-				AND NOT (storage.version = $5 AND storage.read = $6 AND storage.write = $7) -- micro optimization: don't update row unnecessary
 			RETURNING read, write, version, create_time, update_time
 		)
-		(SELECT read, write, version, create_time, update_time from upd)
+		(SELECT read, write, version, create_time, update_time, true AS upsert FROM upd WHERE EXISTS (SELECT 1 FROM upd))
```

Comment:
Thinking about it, `WHERE EXISTS` here is redundant and can be removed:
- if `WHERE` is false, then select  also would not return row if written without `WHERE`.
- if `WHERE` is true, then select without `WHERE` always return row anyway.